File: src/new/my_uio_device.cc
// src/dev/my_uio_device.cc
#include "dev/io_device.hh"
#include "mem/packet.hh"
#include "mem/packet_access.hh"

#include "params/MyUIODevice.hh"
#include "debug/MyUIODevice.hh"
#include "sim/system.hh"

class MyUIODevice : public BasicPioDevice {
public:
    MyUIODevice(const Params *p) : BasicPioDevice(p, p->pio_size) {}

    Tick read(PacketPtr pkt) override {
        // Simulate reading from device
        uint32_t data = 0x12345678;
        pkt->setUintX(data, pkt->getSize());
        pkt->makeAtomicResponse();
        return pioDelay;
    }

    Tick write(PacketPtr pkt) override {
        // Simulate writing to device
        uint32_t data = pkt->getUintX(pkt->getSize());
        DPRINTF(MyUIODevice, "Write: 0x%x\n", data);
        pkt->makeAtomicResponse();
        return pioDelay;
    }

    AddrRangeList getAddrRanges() const override {
        return {RangeSize(pioAddr, pioSize)};
    }
};

File: src/new/MyUIODevice.py
# src/dev/MyUIODevice.py
from m5.params import *
from m5.SimObject import SimObject


class MyUIODevice(SimObject):
    type = "MyUIODevice"
    cxx_header = "new/my_uio_device.cc"
    cxx_class = "MyUIODevice"
    pio_addr = Param.Addr(0x10000000, "Base address")
    pio_size = Param.Addr(0x1000, "Size of address range")
